### PR TITLE
chore(app): rework Application bootstrap

### PR DESCRIPTION
**Pull request by hypeJunction**
Reworks application bootstrap to decouple some services and
eliminate circular dependencies.
Adds application container that holds higher level
application services.
Injects kernel into various services to make it easier to
implement kernel-specific logic in the future.

### CODE REVIEW

File: engine/classes/Elgg/Application/ShutdownHandler.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	 * useful.  Servers will hold pages until processing is done before sending
+	 * them out to the browser.
+	 *
+	 * @see     http://www.php.net/register-shutdown-function
+	 *
+	 * @return void
+	 */
+	public function handleShutdown() {
+		try {
+			if (!$this->application->_services->db) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Application/ShutdownHandler.php]: if (true)

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 2]
```diff
@@ -575,8 +575,6 @@ public function initConfig(Config $config, ServiceProvider $sp) {
 			};
 		}
 
-		$sp->timer->begin([]);
```
Comment (hypeJunction) [inline engine/classes/Elgg/Di/ServiceProvider.php]: Bring back

**Comment by hypeJunction** [comment]
`Application` should only be a proxy for `elgg()`, all business logic has been moved out into dedicated services.
Eventually, `Request`, `ResponseFactory`, `Logger` and anything that use `setClass` should move into the `ApplicationContainer`. Anything that sniffs CLI should move into `Kernel`. 
Ideally `Hooks` and `Events` would reside in the `ApplicationContainer`, and instead of calling handler methods in `Application`, we would subscribe migrations, upgrades etc to internal system events.

**Comment by mrclay** [comment]
What are the roles of Application, ApplicationContainer, Kernel, and Bootstrap? I can see the code moving around but I don't understand why. Which circular deps are removed?

On first glance injecting Application into classes so they can access `->_services` seems worse than just using `_elgg_services()`. We should inject deps, not the service locator.

**Comment by hypeJunction** [comment]
The way I see it, Application is a proxy to call index(), start() and access public services. Currently, it's doing too much, and there is no way to alter behaviours like inject custom exception handler.

If the services were indeed autonomous, we could inject them, but there is too much cross-referencing, and stuff like `elgg_echo()` and `elgg_view()` is all over the place. In an ideal world, the ApplicationContainer would hold global services that can live without the application, but it's quite impossible. For now, the ApplicationContainer provides a way to alter the behavior of the application when it comes to handling exceptions, upgrades, migrations etc.

The Kernel allows us to implement cli and/or web-specific logic in a cleaner way without having to sniff for CLI all the time. For instance, logging behavior is very much dependant on where the application runs, so it would make more sense to inject the kernel into logger and then have kernel specific output. Also, stuff like testing for URL rewrites can't be done in CLI, so just eliminate that logic from controllers altogether.

The Bootstrap is a file loader. You can call it safely to load core libs before you do anything with the application. Loading of core libraries has nothing to do with the application, so it makes sense to extract that logic into a static class that's immune to all of the application resets and instances.

Injecting deps instead of a service locator is an option, but that would have to happen all across the API, and it's just a pain in the ass. We have too many services that rely on each other. We don't do proper unit testing anymore, because all services are dependant on each other.

Circiular depencies were in the BootService, that being a service is expecting ServiceProvide in boot().

Having AppplicationContainer allows us to separte the concerns, and write more readable code. Application currently has methods with 100s of lines of code with so many nested blocks nobody can tell what's happening there.

**Comment by hypeJunction** [comment]
We can also just put stuff from the application container into the service provider, which will make the service provider even more complicated than it already is.

**Comment by hypeJunction** [comment]
Not going to bother. Will just add some workaround for exception handling to test the cli commands.

**Pull request closed without merging**